### PR TITLE
Fix variable declaration

### LINKED ISSUE #1368: Segfault when instantiating an array that references a global for its size
Something changed between commit fc9d8996f415ecbddc2f7f314380f7ff132b09f2 and ca190303cff9bc7e90b5a2c013eaee5e18377cac. The following worked with the former but segfaults with the latter:

```python
from ltypes import i32

SIZE: i32 = i32(1)

def main() -> None:
    xs: i32[SIZE]

main()
```

```shell
$ lpython --show-c main.py
Traceback (most recent call last):
  Binary file "/home/dylon/Workspace/lpython/src/bin/lpython", in _start()
  File "/build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c", line 310, in __libc_start_main()
  File "/home/dylon/Workspace/lpython/src/bin/lpython.cpp", line 1557, in main()
    return emit_c(arg_file, runtime_library_dir, compiler_options);
  File "/home/dylon/Workspace/lpython/src/bin/lpython.cpp", line 285, in (anonymous namespace)::emit_c(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LFortran::CompilerOptions&)
    auto res = LFortran::asr_to_c(al, *asr, diagnostics,
  File "/home/dylon/Workspace/lpython/src/libasr/codegen/asr_to_c.cpp", line 1163, in LFortran::asr_to_c[abi:cxx11](Allocator&, LFortran::ASR::TranslationUnit_t&, LFortran::diag::Diagnostics&, LFortran::Platform&, long)
    v.visit_asr((ASR::asr_t &)asr);
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/asr.h", line 4199, in LFortran::ASR::BaseVisitor<LFortran::ASRToCVisitor>::visit_asr(LFortran::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/asr.h", line 4176, in void LFortran::ASR::visit_asr_t<LFortran::ASRToCVisitor>(LFortran::ASR::asr_t const&, LFortran::ASRToCVisitor&)
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/asr.h", line 4200, in LFortran::ASR::BaseVisitor<LFortran::ASRToCVisitor>::visit_unit(LFortran::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/asr.h", line 3952, in void LFortran::ASR::visit_unit_t<LFortran::ASRToCVisitor>(LFortran::ASR::unit_t const&, LFortran::ASRToCVisitor&)
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/home/dylon/Workspace/lpython/src/libasr/codegen/asr_to_c.cpp", line 602, in LFortran::ASRToCVisitor::visit_TranslationUnit(LFortran::ASR::TranslationUnit_t const&)
    visit_symbol(*sym);
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/asr.h", line 4202, in LFortran::ASR::BaseVisitor<LFortran::ASRToCVisitor>::visit_symbol(LFortran::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/asr.h", line 3962, in void LFortran::ASR::visit_symbol_t<LFortran::ASRToCVisitor>(LFortran::ASR::symbol_t const&, LFortran::ASRToCVisitor&)
    case symbolType::Function: { v.visit_Function((const Function_t &)x); return; }
  File "/home/dylon/Workspace/lpython/src/libasr/../libasr/codegen/asr_to_c_cpp.h", line 441, in LFortran::BaseCCPPVisitor<LFortran::ASRToCVisitor>::visit_Function(LFortran::ASR::Function_t const&)
    if (ASR::is_a<ASR::Variable_t>(*var_sym)) {
  File "/home/dylon/Workspace/lpython/src/libasr/asr.h", line 31, in bool LFortran::ASR::is_a<LFortran::ASR::Variable_t, LFortran::ASR::symbol_t>(LFortran::ASR::symbol_t const&)
    return T::class_type == x.type;
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in killpg()
Segfault: Signal SIGSEGV (segmentation fault) received
```

If I make `SIZE` a local or replace it altogether with an int literal, it succeeds as before:

```python
from ltypes import i32

def main() -> None:
    SIZE: i32 = i32(1)
    xs: i32[SIZE]

main()
```

```shell
$ lpython --show-c main.py
#include <inttypes.h>

#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <lfortran_intrinsics.h>

#define ASSERT(cond)                                                           \
    {                                                                          \
        if (!(cond)) {                                                         \
            printf("%s%s", "ASSERT failed: ", __FILE__);                       \
            printf("%s%s", "\nfunction ", __func__);                           \
            printf("%s%d%s", "(), line number ", __LINE__, " at \n");          \
            printf("%s%s", #cond, "\n");                                       \
            exit(1);                                                           \
        }                                                                      \
    }
#define ASSERT_MSG(cond, msg)                                                  \
    {                                                                          \
        if (!(cond)) {                                                         \
            printf("%s%s", "ASSERT failed: ", __FILE__);                       \
            printf("%s%s", "\nfunction ", __func__);                           \
            printf("%s%d%s", "(), line number ", __LINE__, " at \n");          \
            printf("%s%s", #cond, "\n");                                       \
            printf("%s", "ERROR MESSAGE:\n");                                  \
            printf("%s%s", msg, "\n");                                         \
            exit(1);                                                           \
        }                                                                      \
    }


struct dimension_descriptor
{
    int32_t lower_bound, length;
};

struct i32
{
    int32_t *data;
    struct dimension_descriptor dims[32];
    int32_t n_dims;
    bool is_allocated;
};


// Implementations
void _xx_lcompilers_changed_main_xx()
{
    int32_t SIZE = 1;
    struct i32 xs_value;
    struct i32* xs = &xs_value;
    int32_t *xs_data = (int32_t*) malloc(sizeof(int32_t)*SIZE);
    xs->data = xs_data;
    xs->n_dims = 1;
    xs->dims[0].lower_bound = 0;
    xs->dims[0].length = SIZE;
    SIZE = 1;
}

void _lpython_main_program()
{
    _xx_lcompilers_changed_main_xx();
}

int main(int argc, char* argv[])
{
    _lpython_main_program();
    return 0;
}
```

### PR DESCRIPTION
**Pull request by Smit-create**
Fixes #1368

cc @dylon

### CODE REVIEW

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -308,6 +308,7 @@ RUN(NAME bindc_05   LABELS llvm c
         EXTRAFILES bindc_05b.c)
 RUN(NAME bindc_06   LABELS llvm c
         EXTRAFILES bindc_06b.c)
+RUN(NAME test_global_decl    LABELS llvm c)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:311]: Always test CPython as well:
```suggestion
RUN(NAME test_global_decl    LABELS cpython llvm c)
```

**Comment by Smit-create** [comment]
This is ready for review. Thanks!

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 2]
```diff
@@ -438,7 +438,8 @@ R"(#include <stdio.h>
             std::vector<std::string> var_order = ASRUtils::determine_variable_declaration_order(x.m_symtab);
             for (auto &item : var_order) {
                 ASR::symbol_t* var_sym = x.m_symtab->get_symbol(item);
-                if (ASR::is_a<ASR::Variable_t>(*var_sym)) {
+                // var_sym can be nullptr and present in the parent scope.
+                if (var_sym != nullptr && ASR::is_a<ASR::Variable_t>(*var_sym)) {
```
Comment (certik) [inline src/libasr/codegen/asr_to_c_cpp.h:442]: Something is wrong here. The `determine_variable_declaration_order` accepts the local symbol table and it is supposed to return only variables from it. So `x.m_symtab->get_symbol(item)` should always succeed. Why doesn't it succeed it here? I think this seems like a bug in `determine_variable_declaration_order`.

Comment (Smit-create) [reply src/libasr/codegen/asr_to_c_cpp.h:442]: I'm trying to resolve that in #1378

**Comment by certik** [comment]
It looks like https://github.com/lcompilers/lpython/pull/1378 will fix this.

**Pull request closed without merging**